implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,314 @@
+        };
+
+        WorkerBuilder::with_mailboxes(
+            Self::create_mailboxes(&addresses, decryptor_outgoing_access_control),
+            worker,
+        )
+        .start(context)
+        .await?;
+
+        let decryptor_remote = addresses.decryptor_remote.clone();
+        debug!(
+            "Starting SecureChannel {} at remote: {}",
+            role.clone(),
+            &decryptor_remote
+        );
+        trace!(
+            "SecureChannel {} started for addresses: {:?}",
+            role.clone(),
+            addresses.clone(),
+        );
+
+        if role.is_initiator() {
+            if let Some(timeout) = timeout {
+                callback_waiter.receive_timeout(timeout).await?;
+            }
+            Ok(addresses.encryptor)
+        } else {
+            // wait until the worker is ready to receive messages
+            context.wait_for(addresses.decryptor_remote.clone()).await?;
+            Ok(addresses.decryptor_remote)
+        }
+    }
+
+    /// Return the route for the other party's handshake worker
+    /// That route is set once and for all once we receive a message coming from that worker
+    fn remote_route(&self) -> Route {
+        self.remote_route
+            .clone()
+            .expect("a remote route must have been already set for sending a message")
+    }
+
+    /// Create mailboxes and access rights for the workers involved in the secure channel creation
+    pub(crate) fn create_mailboxes(
+        addresses: &Addresses,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+    ) -> Mailboxes {
+        let remote_mailbox = Mailbox::new(
+            addresses.decryptor_remote.clone(),
+            // Doesn't matter since we check incoming messages cryptographically,
+            // but this may be reduced to allowing only from the transport connection that was used
+            // to create this channel initially
+            Arc::new(AllowAll),
+            // Communicate to the other side of the channel during key exchange
+            Arc::new(AllowAll),
+        );
+        let internal_mailbox = Mailbox::new(
+            addresses.decryptor_internal.clone(),
+            Arc::new(DenyAll),
+            decryptor_outgoing_access_control,
+        );
+        let api_mailbox = Mailbox::new(
+            addresses.decryptor_api.clone(),
+            Arc::new(LocalSourceOnly),
+            Arc::new(LocalOnwardOnly),
+        );
+
+        Mailboxes::new(remote_mailbox, vec![internal_mailbox, api_mailbox])
+    }
+
+    /// Finalize the handshake by creating a `DecryptorWorker` and an `EncryptorWorker`
+    /// Note that `EncryptorWorker` is actually started as an independent worker while
+    /// the `DecryptorWorker` is directly used by this worker to delegate the decryption of messages
+    async fn finalize(
+        &self,
+        context: &Context,
+        handshake_results: HandshakeResults,
+    ) -> Result<DecryptorWorker> {
```

Comment:
I don't understand what is the purpose of `DecryptorWorker` now. It's not a worker at all. Can we refactor it?

Reply:
It's not a worker anymore and it has been renamed to `DecryptorHandler` (there is still a `Decryptor` mirroring the `Encryptor`)